# Go Routines

## errgroup

https://pkg.go.dev/golang.org/x/sync/errgroup

```shell
go get golang.org/x/sync/errgroup
```

### No stop

```go
func main() {
    urls := []string{
        "https://example.com",
        "https://google.com",
        "https://cloudflare.com",
        "_",
        "hello",
        "http://not-exist.com",
        "https://not-exist.com",
    }

    g := &errgroup.Group{}

    for _, url := range urls {
        url := url
        g.Go(func() error {
            return Fetch(url)
        })
    }

    err := g.Wait()
    if err != nil {
        panic(err)
    }
    fmt.Println("‚úÖ Done")
}

func Fetch(url string) error {
    _, err := http.Get(url)
    if err != nil {
        fmt.Println("üî¥ Failed to GET", url)
        return err
    }
    fmt.Println("üü¢ Succeeded to GET", url)
    return nil
}
```

### Stop when error

* Behave like `Promise.all` in JavaScript
* Expected: `apple` `amazon` `reddit` will success
* Expected: `hello` will fail but `cloudflare` `example` `google` will success

```go
func main() {
    var wg = sync.WaitGroup{}
    wg.Add(2)

    go Run(
        context.WithValue(
            context.Background(), "id", "ü¶ä"),
        &wg,
        []string{
            "https://apple.com",
            "https://reddit.com",
            "https://amazon.com",
        },
    )
    go Run(
        context.WithValue(context.Background(), "id", "üêµ"),
        &wg,
        []string{
            "hello",
            "https://example.com",
            "https://google.com",
            "https://cloudflare.com",
        },
    )

    wg.Wait()
}

func Run(ctx context.Context, wg *sync.WaitGroup, urls []string) {
    id := ctx.Value("id")
    fmt.Println(id, "started")
    defer wg.Done()
    g, ctx := errgroup.WithContext(ctx)

    for _, url := range urls {
        url := url
        select {
        case <-ctx.Done():
            return
        default:
            g.Go(func() error {
                return Fetch(ctx, url)
            })
        }
    }

    err := g.Wait()
    if err != nil {
        fmt.Println(id, "‚ùå", err)
        return
    }
    fmt.Println(id, "‚úÖ")
}

func Fetch(ctx context.Context, url string) error {
    id := ctx.Value("id")
    _, err := http.Get(url)
    if err != nil {
        fmt.Println(id, "failed to GET", url)
        return err
    }
    fmt.Println(id, "succeeded to GET", url)
    return nil
}
```

### Stop with context cancellation

* Behave like `AbortController` & `AbortSignal` in JavaScript
* It will cancel all other ongoing operations when one operation returns an error
* Expected: `apple` `amazon` `reddit` will success
* Expected: `hello` will fail then `cloudflare` `example` `google` will be stopped

```go
func main() {
    var wg = sync.WaitGroup{}
    wg.Add(2)

    go Run(
        context.WithValue(
            context.Background(), "id", "ü¶ä"),
        &wg,
        []string{
            "https://apple.com",
            "https://reddit.com",
            "https://amazon.com",
        },
    )
    go Run(
        context.WithValue(context.Background(), "id", "üêµ"),
        &wg,
        []string{
            "hello",
            "https://example.com",
            "https://google.com",
            "https://cloudflare.com",
        },
    )

    wg.Wait()
}

func Run(ctx context.Context, wg *sync.WaitGroup, urls []string) {
    id := ctx.Value("id")
    fmt.Println(id, "started")
    defer wg.Done()
    g, ctx := errgroup.WithContext(ctx)

    for _, url := range urls {
        url := url
        select {
        case <-ctx.Done():
            return
        default:
            g.Go(func() error {
                return Fetch(ctx, url)
            })
        }
    }

    err := g.Wait()
    if err != nil {
        fmt.Println(id, "‚ùå", err)
        return
    }
    fmt.Println(id, "‚úÖ")
}

func Fetch(ctx context.Context, url string) error {
    id := ctx.Value("id")
    req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
    if err != nil {
        fmt.Println(id, "failed to create request for", url)
        return err
    }
    _, err = http.DefaultClient.Do(req)
    if err != nil {
        fmt.Println(id, "failed to GET", url)
        return err
    }
    fmt.Println(id, "succeeded to GET", url)
    return nil
}
```
